### ISSUE TITLE
staticcheck: false positive SA4023: this comparison is always true

### ISSUE THREAD

**Issue by qiuchengxuan** [golangci/golangci-lint#4032]
### Welcome

- [X] Yes, I'm using a binary release within 2 latest major releases. Only such installations are supported.
- [X] Yes, I've searched similar issues on GitHub and didn't find any.
- [X] Yes, I've read the typecheck section of the FAQ (https://golangci-lint.run/usage/faq/#why-do-you-have-typecheck-errors).
- [X] Yes, I've tried with the standalone linter if available (e.g., gocritic, go vet, etc.). (https://golangci-lint.run/usage/linters/)

### Description of the problem

Lint with example file will result with

```
main.go:30:23: SA4023: this comparison is always true (staticcheck)
        if err := w.Close(); err != nil {
                             ^
main.go:21:20: SA4023(related information): (*command-line-arguments.whatever).Close never returns a nil interface value (staticcheck)
func (w *whatever) Close() error {
                   ^
```

which is false positive

### Version of golangci-lint

<details>

```console
$ golangci-lint --version
golangci-lint has version v1.54.1 built with go1.20.4 from (unknown, mod sum: "h1:0qMrH1gkeIBqCZaaAm5Fwq4xys9rO/lJofHfZURIFFk=") on (unknown)
```

</details>

### Configuration

<details>

```console
# paste configuration file or CLI flags here
```

</details>

### Go environment

<details>

```console
$ go version && go env
go version go1.20.4 linux/amd64
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/buyi/.go/pkg/mod"
GOOS="linux"
GOPATH="/home/buyi/.go"
GOROOT="/usr/lib/golang"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/lib/golang/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.20.4"
GCCGO="gccgo"
GOAMD64="v1"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOWORK=""
CGO_CFLAGS="-O2 -g"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-O2 -g"
CGO_FFLAGS="-O2 -g"
CGO_LDFLAGS="-O2 -g"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -Wl,--no-gc-sections -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build1069768467=/tmp/go-build -gno-record-gcc-switches"
```

</details>

### Verbose output of running

<details>

```console
$ golangci-lint cache clean
$ golangci-lint run -v
INFO [config_reader] Config search paths: [./ /home/buyi/test /home/buyi /home /]
INFO [lintersdb] Active 6 linters: [errcheck gosimple govet ineffassign staticcheck unused]
INFO [loader] Go packages loading at mode 575 (types_sizes|imports|name|compiled_files|deps|exports_file|files) took 86.631154ms
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 352.78µs
INFO [linters_context/goanalysis] analyzers took 0s with no stages
INFO [runner] Issues before processing: 3, after processing: 2
INFO [runner] Processors filtering stat (out/in): skip_files: 3/3, identifier_marker: 3/3, exclude: 3/3, nolint: 3/3, uniq_by_line: 2/3, source_code: 2/2, fixer: 2/2, path_prettifier: 3/3, skip_dirs: 3/3, diff: 2/2, max_same_issues: 2/2, severity-rules: 2/2, path_prefixer: 2/2, cgo: 3/3, exclude-rules: 3/3, max_per_file_from_linter: 2/2, max_from_linter: 2/2, sort_results: 2/2, autogenerated_exclude: 3/3, path_shortener: 2/2, filename_unadjuster: 3/3
INFO [runner] processing took 994.241µs with stages: exclude-rules: 362.766µs, nolint: 283.365µs, identifier_marker: 211.215µs, autogenerated_exclude: 40.129µs, source_code: 35.175µs, path_prettifier: 32.757µs, skip_dirs: 9.725µs, max_same_issues: 4.287µs, uniq_by_line: 3.453µs, path_shortener: 2.342µs, cgo: 1.956µs, max_from_linter: 1.94µs, max_per_file_from_linter: 1.18µs, filename_unadjuster: 1.164µs, fixer: 671ns, skip_files: 461ns, exclude: 388ns, sort_results: 366ns, severity-rules: 344ns, diff: 323ns, path_prefixer: 234ns
INFO [runner] linters took 37.886715ms with stages: goanalysis_metalinter: 36.762819ms
main.go:30:23: SA4023: this comparison is always true (staticcheck)
        if err := w.Close(); err != nil {
                             ^
main.go:21:20: SA4023(related information): (*command-line-arguments.whatever).Close never returns a nil interface value (staticcheck)
func (w *whatever) Close() error {
                   ^
INFO File cache stats: 1 entries of total size 505B
INFO Memory: 3 samples, avg is 37.5MB, max is 44.6MB
INFO Execution took 134.075296ms
```

</details>

### A minimal reproducible example or link to a public repository

<details>

```go
package main

import (
        "fmt"
    "io"
    "os"
)

type container[T any] struct {
        t T
}

func (c *container[T]) FuncAny(fn func (t *T) any) any {
        return fn(&c.t)
}

type whatever struct {
        c container[io.Closer]
}

func (w *whatever) Close() error {
        return w.c.FuncAny(func (closer *io.Closer) any{
                return (*closer).Close()
        }).(error)
}

func main() {
        file, _ := os.Open("/tmp/whatever.txt")
        w := whatever{container[io.Closer]{file}}
        if err := w.Close(); err != nil {
                fmt.Println("error")
        }
}
```

</details>

### Validation

- [X] Yes, I've included all information above (version, config, etc.).

**Comment by ldez** [comment]
Hello,

You have checked this box:
> [x] Yes, I've tried with the standalone linter if available (e.g., gocritic, go vet, etc.). (https://golangci-lint.run/usage/linters/)

but the behavior is the same with staticcheck:

```console
$ staticcheck -version
staticcheck 2023.1.4 (v0.4.4)
$ staticcheck ./...                      
foo.go:30:23: this comparison is always true (SA4023)
        foo.go:30:12: the lhs of the comparison is the 1st return value of this function call
        foo.go:21:20: (*github.com/golangci/sandbox.whatever).Close never returns a nil interface value
```

So it's not related to golangci-lint.


Also, your example doesn't compile, so the report is expected:

```console
$ go run ./foo.go                                                                                                                                         
panic: interface conversion: interface is nil, not error

goroutine 1 [running]:
main.(*whatever).Close(0x495f5e?)
        /home/ldez/sources/go/src/github.com/golangci/sandbox/foo.go:24 +0x39
main.main()
        /home/ldez/sources/go/src/github.com/golangci/sandbox/foo.go:30 +0x45
exit status 2
```

**Comment by qiuchengxuan** [comment]
> ```go
> package main
> 
> import (
>         "fmt"
>     "io"
>     "os"
> )
> 
> type container[T any] struct {
>         t T
> }
> 
> func (c *container[T]) FuncAny(fn func (t *T) any) any {
>         return fn(&c.t)
> }
> 
> type whatever struct {
>         c container[io.Closer]
> }
> 
> func (w *whatever) Close() error {
>         return w.c.FuncAny(func (closer *io.Closer) any{
>                 return (*closer).Close()
>         }).(error)
> }
> 
> func main() {
>         file, _ := os.Open("/tmp/whatever.txt")
>         w := whatever{container[io.Closer]{file}}
>         if err := w.Close(); err != nil {
>                 fmt.Println("error")
>         }
> }
> ```

thanks for reply, sorry for occupying your time
